rfcs/2021-09-21-9292-throttle-transform.md

```diff
@@ -0,0 +1,149 @@
+# RFC 9292 - 2021-09-21 - `throttle` transform
+
+This RFC proposes the addition of a new transform that provides a user the ability to control the throughput of specific event streams.
+
+## Context
+
+- [#258](https://github.com/vectordotdev/vector/issues/258)
+
+## Scope
+
+### In scope
+
+- Dropping events (logs and metrics) to rate limit an event stream
+- Rate limit by both bytes (unserialized) and events
+- Exclude events with VRL conditions
+- Specify buckets based on the value of a key in the event
+
+### Out of scope
+
+- Apply backpressure as a form of rate limiting
+
+## Pain
+
+- Users lack the necessary tooling to control throughput which can cause excessive costs and negatively impact downstream services due to increased load
+- Admins cannot set quotas for users/usergroups utilizing Vector
+
+## Proposal
+
+### User Experience
+
+The `throttle` transform can be used to rate limit specific subsets of your event stream to limit load on downstream services or to enforce quotas on users.
+You can enforce rate limits on events or their raw size, as well as excluding events based on a VRL condition to avoid dropping critical logs. Rate limits
+can be applied globally across all logs or by specifying a key to create buckets of events to rate limit more granularly.
+
+### Implementation
+
+The `throttle` transform will leverage the existing [Governor](https://docs.rs/governor/0.3.2/governor/index.html) crate. We will expose the bytes or events
+per second as using facing configuration (or expand to more time frames) and translate that into a `governor::Quota` for internal use. We should not expose
+the `Quota` directly to users to ensure ease of configuration for end users.
+
+Config:
+
+```rust
+pub struct ThrottleConfig {
+    events_per_second: u32,
+    bytes_per_second: u32,
```

Comment:
Are both required, or can you specify one but not the other?

Reply:
I figured one or the other, but theoretically could do both 🤔 it's definitely unclear as is.

Reply:
If it is possible to allow both, I think that would be the most flexible. I'm unsure if `Governor` makes that easy or not.

Reply:
As I'd understand it there'd be two RateLimiter structs created. One for bytes and one for events... maybe applied sequentially?